Src/Web/Web.Shared.Net/RequestTrackingTelemetryModule.cs

```diff
@@ -402,69 +402,100 @@ private class ChildRequestTrackingSuppressionModule
+            /// Using this as a hash-set of current active requests. The second value is not used.
             /// </summary>
-            private static ConcurrentDictionary<string, long> activeRequests = new ConcurrentDictionary<string, long>(System.Environment.ProcessorCount, MAXSIZE);
-            
+            private static ConcurrentDictionary<string, bool> activeRequestsA = new ConcurrentDictionary<string, bool>(System.Environment.ProcessorCount, MAXSIZE);
+            private static ConcurrentDictionary<string, bool> activeRequestsB = new ConcurrentDictionary<string, bool>(System.Environment.ProcessorCount, MAXSIZE);
+
             /// <summary>
-            /// A request must be tracked as Active in order for telemetry to be recorded within OnEndRequest().
+            /// Request will be tagged with an id to identify if it should be logged later.
+            /// </summary>
+            internal void OnBeginRequest_IdRequest(HttpContext context)
+            {
+                this.TagRequest(context);
+            }
+
+            /// <summary>
+            /// OnEndRequest - Should this request be logged?
+            /// Will compare a request id against a hash-set of known requests.
+            /// If this request is not known, add it to hash-set and return true (safe to log).
+            /// If this request is known, return false (do not log twice).
+            /// Additional requests with the same id will return false.
             /// </summary>
-            internal void EnsureActiveRequestIsTracked(HttpContext context)
+            internal bool OnEndRequest_ShouldLog(HttpContext context)
             {
-                // Simplistic cleanup to prevent memory leaks.
-                if (activeRequests.Count >= MAXSIZE)
+                var headers = context.Request.Headers;
+
+                var rootRequestId = headers[HeaderRootRequestId];
+                if (rootRequestId != null)
                 {
-                    int removeCount = 0;
-                    var timeout = DateTime.UtcNow.AddSeconds(TIMEOUTSECONDS).Ticks;
-                    foreach (var x in activeRequests)
+                    if (!this.IsRequestKnown(rootRequestId))
                     {
-                        if (x.Value < timeout)
-                        {
-                            if (activeRequests.TryRemove(x.Key, out var value))
-                            {
-                                removeCount++;
-                            }
-                        }
+                        // doesn't exist add to A and return true;
+                        this.AddRequestToDictionary(rootRequestId);
+                        return true;
                     }
+                }
+
+                return false;
+            }
+
+            /// <summary>
+            /// Tag new requests.
+            /// Transfer Ids to parent requests.
+            /// </summary>
+            private string TagRequest(HttpContext context)
```

Comment:
Do we plan to use return value in the future? If not, this can be safely removed.

Reply:
Thanks, an earlier implementation was populating the tracking dictionary in OnBeginRequest. Removed.